File: solutions/javascript/dfs/keysAndRooms.js
const START_ROOM = 0;

const markReachableNodesAsVisited = (curNode, graph, visited) => {
    // Base cases
    if (visited.has(curNode)) return;

    // Process node
    visited.add(curNode);

    // Recurse on neighbors
    const neighbors = graph[curNode];
    for (const neighbor of neighbors) {
        markReachableNodesAsVisited(neighbor, graph, visited);
    }
};

const canVisitAllRooms = (rooms) => {
    const visited = new Set();

    markReachableNodesAsVisited(START_ROOM, rooms, visited);

    const numReachableRooms = visited.size;
    const numRooms = rooms.length;

    return numReachableRooms === numRooms;
};

File: solutions/python/dfs/keys_and_rooms.py
class Solution:
    def mark_reachable_nodes_as_visited(self, cur_node, graph, visited):
        # Base cases
        if cur_node in visited:
            return

        # Process node
        visited.add(cur_node)

        # Recurse on neighbors
        neighbors = graph[cur_node]

        for neighbor in neighbors:
            self.mark_reachable_nodes_as_visited(neighbor, graph, visited)

    def canVisitAllRooms(self, rooms: List[List[int]]) -> bool:
        visited = set()

        self.mark_reachable_nodes_as_visited(0, rooms, visited)

        num_reachable_rooms = len(visited)
        num_rooms = len(rooms)

        return num_reachable_rooms == num_rooms